compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt

```diff
@@ -219,6 +220,272 @@ internal class UntilHandler(private val context: CommonBackendContext, private v
     }
 }
 
+/** Builds a [HeaderInfo] for progressions built using the `step` extension function. */
+internal class StepHandler(
+    private val context: CommonBackendContext,
+    private val visitor: IrElementVisitor<HeaderInfo?, Nothing?>
+) : ProgressionHandler {
+
+    private val symbols = context.ir.symbols
+
+    override val matcher = SimpleCalleeMatcher {
+        singleArgumentExtension(FqName("kotlin.ranges.step"), symbols.progressionClasses.map { it.typeWith() })
+        parameter(0) { it.type.isInt() || it.type.isLong() }
+    }
+
+    override fun build(expression: IrCall, data: ProgressionType, scopeOwner: IrSymbol): HeaderInfo? =
+        with(context.createIrBuilder(scopeOwner, expression.startOffset, expression.endOffset)) {
+            // Retrieve the HeaderInfo from the underlying progression (if any).
+            val nestedInfo = expression.extensionReceiver!!.accept(visitor, null) as? ProgressionHeaderInfo
+                ?: return null
+
+            val stepArg = expression.getValueArgument(0)!!
+            // We can return the nested info if its step is constant and its absolute value is the same as the step argument. Examples:
+            //
+            //   1..10 step 1               // Nested step is 1, argument is 1. Equivalent to `1..10`.
+            //   10 downTo 1 step 1         // Nested step is -1, argument is 1. Equivalent to `10 downTo 1`.
+            //   10 downTo 1 step 2 step 2  // Nested step is -2, argument is 2. Equivalent to `10 downTo 1 step 2`.
```

Comment:
Is it possible to add next failing test (if they not present yet)
with  ```10 downTo 1 step 2 step -2```  and  ```1..10 step -2 step 2``` ?

Reply:
I have both already; see `compiler/testData/codegen/box/ranges/stepped/expression/downTo/legalStepThenIllegalStep.kt` and `compiler/testData/codegen/box/ranges/stepped/expression/rangeTo/illegalStepThenLegalStep.kt` for all various combinations. Is that what you're referring to?

Reply:
Yes)